File: Conecta_Cusco/src/main/java/com/conectacusco/service/PublicacionService.java
package com.conectacusco.service;

import com.conectacusco.model.PublicacionServicio;
import com.conectacusco.model.OfertaTrabajo;
import com.conectacusco.model.Usuario;
import com.conectacusco.model.CategoriaServicio;
import com.conectacusco.repository.PublicacionServicioRepository;
import com.conectacusco.repository.OfertaTrabajoRepository;
import com.conectacusco.repository.UsuarioRepository;
import com.conectacusco.repository.CategoriaServicioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PublicacionService {

    @Autowired
    private PublicacionServicioRepository publicacionServicioRepository;

    @Autowired
    private OfertaTrabajoRepository ofertaTrabajoRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private CategoriaServicioRepository categoriaServicioRepository;

    // Métodos para PublicacionServicio
    public List<PublicacionServicio> getAllPublicacionesServicio() {
        return publicacionServicioRepository.findAll();
    }

    public Optional<PublicacionServicio> getPublicacionServicioById(Long id) {
        return publicacionServicioRepository.findById(id);
    }

    public PublicacionServicio createPublicacionServicio(PublicacionServicio publicacionServicio, Long userId, Integer categoriaId) {
        Usuario usuario = usuarioRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado con ID: " + userId));
        CategoriaServicio categoria = categoriaServicioRepository.findById(categoriaId)
                .orElseThrow(() -> new RuntimeException("Categoría no encontrada con ID: " + categoriaId));
        publicacionServicio.setUsuario(usuario);
        publicacionServicio.setCategoria(categoria);
        return publicacionServicioRepository.save(publicacionServicio);
    }

    public PublicacionServicio updatePublicacionServicio(Long id, PublicacionServicio publicacionDetails) {
        PublicacionServicio publicacion = publicacionServicioRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Publicación de servicio no encontrada con ID: " + id));
        publicacion.setTitulo(publicacionDetails.getTitulo());
        publicacion.setDescripcion(publicacionDetails.getDescripcion());
        publicacion.setPrecio(publicacionDetails.getPrecio());
        publicacion.setUbicacion(publicacionDetails.getUbicacion());
        // No actualizamos usuario ni categoría directamente aquí, para cambios se debería manejar lógica específica
        return publicacionServicioRepository.save(publicacion);
    }

    public void deletePublicacionServicio(Long id) {
        publicacionServicioRepository.deleteById(id);
    }

    public List<PublicacionServicio> searchPublicacionesServicio(String query) {
        return publicacionServicioRepository.findByTituloContainingIgnoreCaseOrDescripcionContainingIgnoreCase(query, query);
    }

    public List<PublicacionServicio> getPublicacionesServicioByCategoria(String categoryName) {
        return publicacionServicioRepository.findByCategoriaNombreContainingIgnoreCase(categoryName);
    }

    // Métodos para OfertaTrabajo
    public List<OfertaTrabajo> getAllOfertasTrabajo() {
        return ofertaTrabajoRepository.findAll();
    }

    public Optional<OfertaTrabajo> getOfertaTrabajoById(Long id) {
        return ofertaTrabajoRepository.findById(id);
    }

    public OfertaTrabajo createOfertaTrabajo(OfertaTrabajo ofertaTrabajo, Long userId, Integer categoriaId) {
        Usuario usuario = usuarioRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado con ID: " + userId));
        CategoriaServicio categoria = categoriaServicioRepository.findById(categoriaId)
                .orElseThrow(() -> new RuntimeException("Categoría no encontrada con ID: " + categoriaId));
        ofertaTrabajo.setUsuario(usuario);
        ofertaTrabajo.setCategoria(categoria);
        return ofertaTrabajoRepository.save(ofertaTrabajo);
    }

    public OfertaTrabajo updateOfertaTrabajo(Long id, OfertaTrabajo ofertaDetails) {
        OfertaTrabajo oferta = ofertaTrabajoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Oferta de trabajo no encontrada con ID: " + id));
        oferta.setTitulo(ofertaDetails.getTitulo());
        oferta.setDescripcion(ofertaDetails.getDescripcion());
        oferta.setPresupuesto(ofertaDetails.getPresupuesto());
        oferta.setUbicacion(ofertaDetails.getUbicacion());
        return ofertaTrabajoRepository.save(oferta);
    }

    public void deleteOfertaTrabajo(Long id) {
        ofertaTrabajoRepository.deleteById(id);
    }

    public List<OfertaTrabajo> searchOfertasTrabajo(String query) {
        return ofertaTrabajoRepository.findByTituloContainingIgnoreCaseOrDescripcionContainingIgnoreCase(query, query);
    }

    public List<OfertaTrabajo> getOfertasTrabajoByCategoria(String categoryName) {
        return ofertaTrabajoRepository.findByCategoriaNombreContainingIgnoreCase(categoryName);
    }
}

File: Conecta_Cusco_Frontend/src/app/services/publicacion.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PublicacionServicio } from '../models/publicacion.model';
import { OfertaTrabajo } from '../models/oferta.model';

const API_URL = 'http://localhost:8080/api/publicaciones/';

@Injectable({
  providedIn: 'root'
})
export class PublicacionService {

  constructor(private http: HttpClient) { }

  // Métodos para PublicacionServicio
  getAllPublicacionesServicio(): Observable<PublicacionServicio[]> {
    return this.http.get<PublicacionServicio[]>(API_URL + 'servicios');
  }

  getPublicacionServicioById(id: number): Observable<PublicacionServicio> {
    return this.http.get<PublicacionServicio>(API_URL + 'servicios/' + id);
  }

  createPublicacionServicio(publicacion: any, userId: number, categoriaId: number): Observable<PublicacionServicio> {
    return this.http.post<PublicacionServicio>(`${API_URL}servicios/crear/${userId}/${categoriaId}`, publicacion);
  }

  updatePublicacionServicio(id: number, publicacion: PublicacionServicio): Observable<PublicacionServicio> {
    return this.http.put<PublicacionServicio>(API_URL + 'servicios/actualizar/' + id, publicacion);
  }

  deletePublicacionServicio(id: number): Observable<any> {
    return this.http.delete(API_URL + 'servicios/eliminar/' + id);
  }

  searchPublicacionesServicio(query: string): Observable<PublicacionServicio[]> {
    return this.http.get<PublicacionServicio[]>(`${API_URL}servicios/buscar?query=${query}`);
  }

  getPublicacionesServicioByCategoria(category: string): Observable<PublicacionServicio[]> {
    return this.http.get<PublicacionServicio[]>(`${API_URL}servicios/categoria?category=${category}`);
  }

  // Métodos para OfertaTrabajo
  getAllOfertasTrabajo(): Observable<OfertaTrabajo[]> {
    return this.http.get<OfertaTrabajo[]>(API_URL + 'ofertas');
  }

  getOfertaTrabajoById(id: number): Observable<OfertaTrabajo> {
    return this.http.get<OfertaTrabajo>(API_URL + 'ofertas/' + id);
  }

  createOfertaTrabajo(oferta: any, userId: number, categoriaId: number): Observable<OfertaTrabajo> {
    return this.http.post<OfertaTrabajo>(`${API_URL}ofertas/crear/${userId}/${categoriaId}`, oferta);
  }

  updateOfertaTrabajo(id: number, oferta: OfertaTrabajo): Observable<OfertaTrabajo> {
    return this.http.put<OfertaTrabajo>(API_URL + 'ofertas/actualizar/' + id, oferta);
  }

  deleteOfertaTrabajo(id: number): Observable<any> {
    return this.http.delete(API_URL + 'ofertas/eliminar/' + id);
  }

  searchOfertasTrabajo(query: string): Observable<OfertaTrabajo[]> {
    return this.http.get<OfertaTrabajo[]>(`${API_URL}ofertas/buscar?query=${query}`);
  }

  getOfertasTrabajoByCategoria(category: string): Observable<OfertaTrabajo[]> {
    return this.http.get<OfertaTrabajo[]>(`${API_URL}ofertas/categoria?category=${category}`);
  }
}